File: apps/frontend/lib/countries.ts
import type { Lang } from "./i18n";

export const COUNTRIES = [
  { code: "TH", label: "Thailand", labelTh: "ไทย" },
  { code: "KR", label: "South Korea", labelTh: "เกาหลีใต้" },
  { code: "JP", label: "Japan", labelTh: "ญี่ปุ่น" },
  { code: "US", label: "United States", labelTh: "สหรัฐอเมริกา" },
  { code: "GB", label: "United Kingdom", labelTh: "สหราชอาณาจักร" },
] as const;

export type CountryCode = (typeof COUNTRIES)[number]["code"];

const COUNTRY_CODE_SET = new Set<string>(COUNTRIES.map((c) => c.code));

export function isValidCountry(code: string): code is CountryCode {
  return COUNTRY_CODE_SET.has(code);
}

export function countryLabel(code: string, lang: Lang = "en"): string {
  const country = COUNTRIES.find((c) => c.code === code);
  if (!country) return code;
  return lang === "th" ? country.labelTh : country.label;
}

export const DEFAULT_COUNTRY: CountryCode = "TH";

export function detectCountryFromAcceptLanguage(acceptLanguage: string | null): CountryCode {
  if (!acceptLanguage) return DEFAULT_COUNTRY;

  const tags = acceptLanguage
    .split(",")
    .map((part) => part.split(";")[0]?.trim().toLowerCase())
    .filter((tag): tag is string => Boolean(tag));

  for (const tag of tags) {
    if (tag.startsWith("th")) return "TH";
    if (tag.startsWith("ko")) return "KR";
    if (tag.startsWith("ja")) return "JP";
    if (tag.startsWith("en-gb") || tag === "en-uk") return "GB";
    if (tag.startsWith("en")) return "US";
  }

  return DEFAULT_COUNTRY;
}

File: apps/mobile/lib/constants/countries.dart
import 'dart:ui';

import '../i18n/lang.dart';

class Country {
  final String code;
  final String label;
  final String labelTh;

  const Country(this.code, this.label, this.labelTh);
}

const List<Country> kCountries = [
  Country('TH', 'Thailand', 'ไทย'),
  Country('KR', 'South Korea', 'เกาหลีใต้'),
  Country('JP', 'Japan', 'ญี่ปุ่น'),
  Country('US', 'United States', 'สหรัฐอเมริกา'),
  Country('GB', 'United Kingdom', 'สหราชอาณาจักร'),
];

const String kDefaultCountry = 'TH';

bool isValidCountry(String code) {
  return kCountries.any((c) => c.code == code);
}

String countryLabel(String code, AppLang lang) {
  for (final country in kCountries) {
    if (country.code == code) {
      return lang == AppLang.th ? country.labelTh : country.label;
    }
  }
  return code;
}

String detectCountryFromLocale(Locale locale) {
  final language = locale.languageCode.toLowerCase();
  final region = (locale.countryCode ?? '').toUpperCase();

  if (language == 'th') return 'TH';
  if (language == 'ko') return 'KR';
  if (language == 'ja') return 'JP';
  if (language == 'en' && region == 'GB') return 'GB';
  if (language == 'en') return 'US';

  return kDefaultCountry;
}
